Testing click() gives "RuntimeError: dictionary changed size during iteration"

### Discussed in https://github.com/zauberzeug/nicegui/discussions/5935

<div type='discussions-op-text'>

<sup>Originally posted by **samuller** April  6, 2026</sup>
### First Check

- [X] I added a very descriptive title here.
- [X] I have checked the [FAQs](/zauberzeug/nicegui/wiki/FAQs) for a similar question and didn't find it.
- [X] I used the GitHub search to find a similar [question](https://github.com/zauberzeug/nicegui/discussions/categories/q-a) and didn't find it.
- [X] I searched the NiceGUI documentation, with the integrated search.
- [X] I already searched online "How to X in NiceGUI" and didn't find any information.

### Example Code

```python
"""Minimal reproducible example for:
    RuntimeError: dictionary changed size during iteration

The bug:
- A button lives inside a container.
- Its click handler calls container.clear(), which deletes the button.
- NiceGUI's testing User.click() iterates `element._event_listeners.values()` to
  fire the handler.  Deleting the element mid-iteration calls _handle_delete() →
  _event_listeners.clear(), mutating the dict while it is being iterated.

Run with:
    pytest dict_iter_bug.py
"""

from nicegui import ui
from nicegui.testing import User

pytest_plugins = ["nicegui.testing.user_plugin"]


def setup():
    @ui.page("/demo")
    def page():
        container = ui.row()

        def rebuild():
            container.clear()
            with container:
                # Button lives INSIDE the container it clears — that's the bug trigger.
                ui.button("click me", on_click=rebuild).mark("the_button")

        rebuild()


async def test_click_causes_self_destruction(user: User):
    setup()
    await user.open("/demo")
    # The button's handler calls container.clear(), which destroys the button while
    # NiceGUI testing is still iterating its _event_listeners dict →
    # RuntimeError: dictionary changed size during iteration
    user.find("the_button").click()
    await user.should_see("click me")
```


### Description

The above code replicates the issue when testing a button whose `click()` behavior causes itself to be re-created. I get this RuntimeError only during tests, but otherwise my code works perfectly fine:
```python
                    ...
                    if target_key in expanded_set:
                        expanded_set.remove(target_key)
                    else:
                        expanded_set.add(target_key)
                    element.props['expanded'] = list(expanded_set)
                    element.update()
                    return self
    
>               for listener in element._event_listeners.values():  # pylint: disable=protected-access
E               RuntimeError: dictionary changed size during iteration

.venv/lib/python3.12/site-packages/nicegui/testing/user_interaction.py:137: RuntimeError
```
I know I could try to hide/show the button rather than delete/re-create it, but for my full code the current approach works cleanly and it's only tests that give a problem.

Is this a bug in the test framework or is this a coding pattern I should be avoiding?

### NiceGUI Version

3.8.0

### Python Version

3.12.3

### Browser

Firefox

### Operating System

Linux

### Additional Context

_No response_</div>

--- a/nicegui/testing/user_interaction.py
+++ b/nicegui/testing/user_interaction.py
@@ -46,11 +46,13 @@ class UserInteraction(Generic[T]):
                             element.value = option
                             break
 
-                for listener in list(element._event_listeners.values()):  # pylint: disable=protected-access
+                for listener in element._event_listeners.values():  # pylint: disable=protected-access
                     if listener.type != event:
                         continue
                     event_arguments = events.GenericEventArguments(sender=element, client=self.user.client, args=args)
                     events.handle_event(listener.handler, event_arguments)
+                    if element.is_deleted:
+                        break
         return self
 
     def type(self, text: str) -> Self:
@@ -139,12 +141,14 @@ class UserInteraction(Generic[T]):
                     element.update()
                     return self
 
-                for listener in list(element._event_listeners.values()):  # pylint: disable=protected-access
+                for listener in element._event_listeners.values():  # pylint: disable=protected-access
                     if listener.element_id != element.id:
                         continue
                     args = not element.value if isinstance(element, (ui.checkbox, ui.switch)) else None
                     event_arguments = events.GenericEventArguments(sender=element, client=self.user.client, args=args)
                     events.handle_event(listener.handler, event_arguments)
+                    if element.is_deleted:
+                        break
         return self
 
     def clear(self) -> Self:

--- a/nicegui/testing/user_interaction.py
+++ b/nicegui/testing/user_interaction.py
@@ -46,7 +46,7 @@ class UserInteraction(Generic[T]):
                             element.value = option
                             break
 
-                for listener in element._event_listeners.values():  # pylint: disable=protected-access
+                for listener in list(element._event_listeners.values()):  # pylint: disable=protected-access
                     if listener.type != event:
                         continue
                     event_arguments = events.GenericEventArguments(sender=element, client=self.user.client, args=args)
@@ -139,7 +139,7 @@ class UserInteraction(Generic[T]):
                     element.update()
                     return self
 
-                for listener in element._event_listeners.values():  # pylint: disable=protected-access
+                for listener in list(element._event_listeners.values()):  # pylint: disable=protected-access
                     if listener.element_id != element.id:
                         continue
                     args = not element.value if isinstance(element, (ui.checkbox, ui.switch)) else None

--- a/tests/test_user_simulation.py
+++ b/tests/test_user_simulation.py
@@ -893,10 +893,15 @@ async def test_clearing_container_with_button_inside(user: User) -> None:
 
         def rebuild():
             with container.clear():
-                ui.button('click me', on_click=rebuild)
+                ui.button('click me') \
+                    .on('click', lambda: ui.notify('First handler')) \
+                    .on('click', rebuild) \
+                    .on('click', lambda: ui.notify('Last handler'))
 
         rebuild()
 
     await user.open('/')
     user.find('click me').click()
+    await user.should_see('First handler')
     await user.should_see('click me')
+    await user.should_not_see('Last handler')

--- a/tests/test_user_simulation.py
+++ b/tests/test_user_simulation.py
@@ -884,3 +884,19 @@ async def test_switching_tabs_wrapped_in_row(user: User) -> None:
     await user.open('/')
     user.find('A').click()
     await user.should_see('Switching to A')
+
+
+async def test_clearing_container_with_button_inside(user: User) -> None:
+    @ui.page('/')
+    def page():
+        container = ui.row()
+
+        def rebuild():
+            with container.clear():
+                ui.button('click me', on_click=rebuild)
+
+        rebuild()
+
+    await user.open('/')
+    user.find('click me').click()
+    await user.should_see('click me')

